distributed/spill.py

```diff
@@ -49,28 +59,56 @@ def disk(self) -> Mapping[Hashable, Any]:
         """
         return self.slow
 
-    @staticmethod
-    def _weight(key: Hashable, value: Any) -> int:
-        return safe_sizeof(value)
-
-    def _on_evict(self, key: Hashable, value: Any) -> None:
-        b = safe_sizeof(value)
-        self.spilled_by_key[key] = b
-        self.spilled_total += b
-
-    def _on_retrieve(self, key: Hashable, value: Any) -> None:
-        self.spilled_total -= self.spilled_by_key.pop(key)
-
-    def __setitem__(self, key: Hashable, value: Any) -> None:
-        self.spilled_total -= self.spilled_by_key.pop(key, 0)
-        super().__setitem__(key, value)
-        if key in self.slow:
-            # value is individually larger than target so it went directly to slow.
-            # _on_evict was not called.
-            b = safe_sizeof(value)
-            self.spilled_by_key[key] = b
-            self.spilled_total += b
-
-    def __delitem__(self, key: Hashable) -> None:
-        self.spilled_total -= self.spilled_by_key.pop(key, 0)
+    @property
+    def spilled_total(self) -> int:
+        return self.slow.total_weight
+
+
+def _in_memory_weight(key: Hashable, value: Any) -> int:
+    return safe_sizeof(value)
+
+
+class MaxSpillExceeded(Exception):
+    pass
+
+
+class Slow(zict.Func):
+    weight_by_key: dict[Hashable, int]
+    total_weight: int
+
+    def __init__(self, spill_directory: str, max_weight: int | Literal[False] = False):
+        super().__init__(
+            partial(serialize_bytelist, on_error="raise"),
+            deserialize_bytes,
+            zict.File(spill_directory),
+        )
+        self.max_weight = max_weight
+        self.weight_by_key = {}
+        self.total_weight = 0
+
+    def __setitem__(self, key, value):
+        pickled = self.dump(value)
+        pickled_size = sum(len(frame) for frame in pickled)
+
+        if self.max_weight and self.total_weight + pickled_size > self.max_weight:
+            # TODO don't spam the log file with hundreds of messages per second
+            logger.warning(
+                "Spill file on disk reached capacity; keeping data in memory"
+            )
+            # Stop callbacks and ensure that the key ends up in SpillBuffer.fast
+            self.total_weight -= self.weight_by_key.pop(
+                key, 0
+            )  # isn't this taken care on delitem
+            self.d.pop(key, None)
+            raise MaxSpillExceeded()
+        print(f"{key = }")
+        print(f"{self.weight_by_key.get(key, 0) = }")
+        self.total_weight += pickled_size  # - self.weight_by_key.get(key, 0) seem to be not having any effect when overwriting a key because if key in slow in the buffer we delete it
+        self.weight_by_key[key] = pickled_size
+        self.d[
+            key
+        ] = pickled  # this goes under a set item on File which does something to pickle
+
+    def __delitem__(self, key):
         super().__delitem__(key)
+        self.total_weight -= self.weight_by_key.pop(key, 0)
```

Comment:
if Func.__delitem__ doesn't raise, then you are guaranteed to always have the key in weight_by_key
```suggestion
        self.total_weight -= self.weight_by_key.pop(key)
```